### PR TITLE
DAOS-5682 csum: Bug fixes for rebuild with csum enabled

### PR DESCRIPTION
**Pull request by ryon-jensen**
- Fixed several places in the rebuild process where
  punched record and key checksums weren't processed
  correctly.
- Added some notes to the data_integrity wiki page about
  some key points in the code for checksums.
- Fixed an issue on the server when adding csums to
  the the iod on a fetch. Previously the wrong index
  was used.

Signed-off-by: Ryon Jensen <ryon.jensen@intel.com>

### CODE REVIEW

File: doc/overview/data_integrity.md [hunk 1]
```diff
@@ -179,4 +179,39 @@ Unit and functional testing is performed at many layers.
...
 ./daos_server -i --csum_type crc64
-```
\ No newline at end of file
+```
+
+# Life of a checksum (WIP)
+## Rebuild
+- migrate_one_insert - mrone.iods_csums is allocated, iods_csums copied from
+  dss_enum_unpack_io. Memory reference removed from iods_csums.data. It will be
+  freed in migrate_one_destory
```
Comment (jolivier23) [inline doc/overview/data_integrity.md:188]: s/destory/destroy

**Review by jolivier23 — APPROVED** [review]

**Comment by ryon-jensen** [comment]
rebuild test - DAOS-5784, DAOS-5305
dmg pool query test - DAOS-5800
csum_error_logging.py - DAOS-5745
Still investigating the osa_online_reintegration test failure

**Review by rpadma2 — APPROVED** [review]
Changes looks fine. Are we sure online reintegration failure is not related to this change? I would run that particular test case to make sure this PR didn't introduce any new regressions.

**Review by jgmoore-or — APPROVED** [review]

**Comment by ryon-jensen** [comment]
Ravi discovered the osa online reintegration test failure on master as well so created a ticket for it. This should be ready for gatekeeper to land.

**Pull request merged**